core/src/main/java/org/apache/iceberg/stats/StatsWriter.java

```diff
@@ -0,0 +1,145 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.stats;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import javax.annotation.Nullable;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.io.PositionOutputStream;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+public class StatsWriter implements Closeable {
+
+  private final PositionOutputStream outputStream;
+
+  private final Map<String, String> properties = Maps.newHashMap();
+  private final List<BlobMetadata> blobs = Lists.newArrayList();
+  private boolean compressFooter = false; // TODO compress footer by default when LZ4 support is added
+
+  private boolean headerWritten;
+  private boolean finished;
+  private Optional<Integer> footerSize = Optional.empty();
+
+  public StatsWriter(OutputFile outputFile) {
+    Preconditions.checkNotNull(outputFile, "outputFile is null");
+    this.outputStream = outputFile.create();
+  }
+
+  public void addFileProperty(String name, String value) {
+    Preconditions.checkNotNull(name, "name is null");
+    Preconditions.checkNotNull(value, "value is null");
+
+    if (properties.putIfAbsent(name, value) != null) {
+      throw new IllegalStateException(String.format("Property '%s' already set", name));
+    }
+  }
+  public void append(
+      String type,
+      Set<Integer> columnsCovered,
+      ByteBuffer blobData,
+      Optional<StatsCompressionCodec> compression) throws IOException {
+    checkNotFinished();
+    writeHeaderIfNeeded();
+
+    Preconditions.checkNotNull(type, "type is null");
+    long fileOffset = outputStream.getPos();
+    ByteBuffer data;
+    data = compression.map(codec -> StatsFormat.compressBlob(codec, blobData))
+        .orElse(blobData);
+    int length = data.remaining();
+    Channels.newChannel(outputStream).write(data);
+    @Nullable String codecName = compression.map(StatsCompressionCodec::getCodecName).orElse(null);
+    blobs.add(new BlobMetadata(type, columnsCovered, fileOffset, length, codecName));
+  }
+
+  public void setCompressFooter(boolean compressFooter) {
+    this.compressFooter = compressFooter;
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (!finished) {
+      finish();
+    }
+
+    outputStream.close();
+  }
+
+  private void writeHeaderIfNeeded() throws IOException {
+    if (headerWritten) {
+      return;
+    }
+
+    outputStream.write(StatsFormat.getMagic());
+    headerWritten = true;
```

Comment:
When modifying state by setting an instance field, we like to use `this.` so it is obvious that it isn't a local variable.